caffe2/core/tensor_impl.h

```diff
@@ -785,68 +789,45 @@ class CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
   TypeMeta data_type_;
 
  private:
+  TIndex compute_numel() const {
+    TIndex n = 1;
+    for (auto s : sizes()) {
+      n *= s;
+    }
+    return n;
+  }
+
   template <
       typename T,
       typename = typename std::enable_if<std::is_integral<T>::value>::type>
   bool SetDims(const std::vector<T>& src) {
-    auto old_numel = numel_;
-    dims_.resize(src.size());
-    TIndex new_numel = 1;
     for (size_t i = 0; i < src.size(); ++i) {
-      new_numel *= src[i];
-      dims_[i] = src[i];
+      sizes_[i] = src[i];
     }
-    numel_ = new_numel;
+    auto old_numel = numel_;
+    numel_ = compute_numel();
     return numel_ != old_numel;
   }
 
   bool SetDims() {
-    auto old_numel = numel_;
-    dims_.resize(0);
-    numel_ = 1;
-    return numel_ != old_numel;
+    return SetDims<TIndex>({});
   }
 
-  // TODO(jiayq): maybe rewrite the following functions with initializer list.
-  // NVCC does not play well with initializer lists last time, but worth
-  // another shot.
   bool SetDims(const TIndex d0) {
```

Comment:
Can't you just get rid of these functions entirely? They're private...

Reply:
They're used within Resize to support different semantics.